trinity/extensibility/asyncio.py

```diff
@@ -28,35 +31,55 @@ async def _run_in_process(
     ) -> TReturn:
         return await run_in_process(async_fn, *args, subprocess_kwargs=subprocess_kwargs)
 
+    # FIXME: Disabled by default
+    async def _loop_monitoring_task(self) -> None:
+        while True:
+            timer = Timer()
+            await asyncio.sleep(self.loop_monitoring_wakeup_interval)
+            variance = timer.elapsed - self.loop_monitoring_wakeup_interval
+            if variance > self.loop_monitoring_max_variance:
+                pending_tasks = len([task for task in asyncio.all_tasks() if not task.done()])
+                self.logger.warning(
+                    "Event loop blocked or overloaded: variance=%f, tasks=%d",
```

Comment:
After sitting on this a bit, I think we want one more metric here and some extra logic about when to log this warning.

For the logging, I think we want to reduce how often this gets dumped to warning since it would effectively be dumped every 2 seconds if our loop is consistently blocked.   Maybe update with some code to log to `DEBUG` if a message was logged to warning with the last few minutes.

I also think we might like a little extra information here on how *often* this is occuring.  Basically, something like a percentage of the time that the loop ends up blocked.  We could use `p2p.stats.ema.EMA` and record a `0` when the loop doesn't block and a `100` when the loop does block.  This would gives us a percentage to include in this message which gives us information that not only is the loop being blocked, but how often.

Reply:
IMO, the log would be a little easier to read with only 3 digits and a unit. Also, "variance" in this context has a name collision with the square of the standard deviation. Maybe "lag" or "delay"? Something like:
```suggestion
                    "Event loop blocked or overloaded: delay=%.3fs, tasks=%d",
```


> Maybe update with some code to log to `DEBUG` if a message was logged to warning with the last few minutes.

Yeah, the `TokenBucket` is used for this in a couple places, and has worked well. We can have a capacity of 5ish and a slow refill rate (~1 per 5 minute). Then you several logs in a row at the beginning to show you that it's really blocked, but then the logs quiet down for long runs. (unless the event loop cleared up and then got blocked again)

That said, there's clearly really useful stuff coming out of these logs already. So I'd definitely prioritize getting this merged over tinkering for too long with the `TokenBucket`.

Reply:
Also, this is still running as an unsupervised background task and we have no tests for it, so if I introduce bugs it'd stop working and we wouldn't find out easily. I'll merge it with @carver's suggestion, then work on a test and try to address the FIXME so that we terminate the component in case the monitoring task terminates early. Then I can work on @pipermerriam's suggestions knowing that at least regressions that cause it to crash will be caught by tests